send dhcpv6 release message on stop

Strangelovian opened the pull request:
Have the DHCPv6 client to send DHCPv6 release message when it stops.
https://github.com/systemd/systemd/issues/19728

Use case: when using systemd as a home router, this is useful to release both DHCPv4 and DHCPv6 leases, when rebooting several times in a row. Otherwise, one can be locked out of IPv6, if the ISP refuse to issue more than a few simultaneously active DHCPv6 leases.

Analogous to the existing DHCPv4 client parameter: same default value true, the DHCPv6 release is sent by default.

Tested successfully when back ported on a debian 11 intel x64 home router.

Strangelovian commented:
With the clang build the C99 bool type is refused. I guess it's expected with testing compat of older C standards.
Is it expected? It's passing with the GCC build.
```
 145/1090 systemd:headers / cc-sd-dhcp6-client.h_c-ansi                        FAIL           0.12s (exit status 1)
--- command ---
22:42:01 /usr/bin/env clang -c -x c -ansi -Wall -Wextra -Werror -pedantic -Wno-long-long -Wno-variadic-macros -include /home/chris/src/gh/systemd/src/systemd/sd-dhcp6-client.h -o/dev/null /dev/null
--- stderr ---
In file included from <built-in>:1:
/home/chris/src/gh/systemd/src/systemd/sd-dhcp6-client.h:270:38: error: '_Bool' is a C99 extension [-Werror,-Wc99-extensions]
                                     bool enable);
```

yuwata commented on man/systemd.network.xml:
```diff
@@ -2305,6 +2305,15 @@ allow my_server_t localnet_peer_t:peer recv;</programlisting>
+      <varlistentry>
+        <term><varname>SendRelease=</varname></term>
+        <listitem>
+          <para>When true, the DHCPv6 client sends a DHCP release packet when it stops. Defaults to
```
Please clarify what kind of value is taken: `Takes a boolean value.`

yuwata replied:
Or, maybe, you can simply append the setting after `NetLabel=`.

yuwata commented on src/libsystemd-network/sd-dhcp6-client.c:
```diff
@@ -498,6 +498,14 @@ int sd_dhcp6_client_set_rapid_commit(sd_dhcp6_client *client, int enable) {
         return 0;
 }
 
+int sd_dhcp6_client_set_send_release(sd_dhcp6_client *client, bool enable) {
```
As you already noticed, please use `int enable` even if the internal variable `client->send_release` is `bool`.

yuwata commented on src/libsystemd-network/sd-dhcp6-client.c:
```diff
@@ -1331,6 +1343,8 @@ int sd_dhcp6_client_stop(sd_dhcp6_client *client) {
 
         client_stop(client, SD_DHCP6_CLIENT_EVENT_STOP);
+        if (client->send_release)
+                dhcp6_client_send_message(client);
```
Missing error handling. I think it is OK to ignore the failure, but please log about that.

yuwata commented on src/libsystemd-network/test-dhcp6-client.c:
```diff
@@ -974,6 +974,9 @@ int dhcp6_network_send_udp_socket(int s, struct in6_addr *a, const void *packet,
                 break;
 
         case 3:
+                break;
```
Please verify the packet.

yuwata commented on src/network/networkd-dhcp6.c:
```diff
@@ -707,6 +707,12 @@ static int dhcp6_configure(Link *link) {
+        r = sd_dhcp6_client_set_send_release(client, link->network->dhcp6_send_release);
+        if (r < 0)
+                return log_link_debug_errno(link, r,
+                                            "DHCPv6 CLIENT: Failed to %s send release: %m",
```
`send release` -> `sending release message on stop` ??

yuwata commented on src/systemd/sd-dhcp6-client.h:
```diff
@@ -263,7 +264,10 @@ int sd_dhcp6_client_set_address_request(sd_dhcp6_client *client,
                                         int request);
 int sd_dhcp6_client_add_vendor_option(sd_dhcp6_client *client,
                                       sd_dhcp6_option *v);
-int sd_dhcp6_client_set_rapid_commit(sd_dhcp6_client *client, int enable);
+int sd_dhcp6_client_set_rapid_commit(sd_dhcp6_client *client,
+                                     int enable);
```
It is not necessary to wrap lines too frequently. Please keep the line as is.

yuwata commented on src/systemd/sd-dhcp6-client.h:
```diff
@@ -263,7 +264,10 @@ int sd_dhcp6_client_set_address_request(sd_dhcp6_client *client,
+int sd_dhcp6_client_set_rapid_commit(sd_dhcp6_client *client,
+                                     int enable);
+int sd_dhcp6_client_set_send_release(sd_dhcp6_client *client,
+                                     bool enable);
```
`int enable`

yuwata commented on src/systemd/sd-dhcp6-client.h:
```diff
@@ -22,6 +22,7 @@
 #include <inttypes.h>
 #include <net/ethernet.h>
 #include <sys/types.h>
+#include <stdbool.h>
```
Please drop it.

yuwata reviewed (changes requested):
Could you provide the reference to the RFC about the release message?
In the release message, what kind of attributes should be included??

Strangelovian commented:
@yuwata Thanks for reviewing. I addressed your remarks, please let me know.

> Could you provide the reference to the RFC about the release message? In the release message, what kind of attributes should be included??

Yes included some explanations in the test code here https://github.com/systemd/systemd/blob/c497609372d74a76364afd8fa05b43710015def0/src/libsystemd-network/test-dhcp6-client.c#L605

Strangelovian commented:
Example of DHCP6 release sent with this change:
<img width="540" alt="Capture d’écran 2023-01-15 à 19 46 57" src="https://user-images.githubusercontent.com/3154712/212561052-588aeb3b-7553-4543-a5e5-ba0d711f83e6.png">
Reply from a dnsmaq server:
<img width="541" alt="Capture d’écran 2023-01-15 à 19 53 44" src="https://user-images.githubusercontent.com/3154712/212561255-e826d5b3-06ee-4272-8662-1ad6c9131a51.png">

yuwata commented on src/libsystemd-network/dhcp6-protocol.c:
```diff
@@ -11,6 +11,7 @@ static const char * const dhcp6_state_table[_DHCP6_STATE_MAX] = {
         [DHCP6_STATE_BOUND]               = "bound",
         [DHCP6_STATE_RENEW]               = "renew",
         [DHCP6_STATE_REBIND]              = "rebind",
+        [DHCP6_STATE_STOPPING]             = "stopping",
```
Spurious alignment.

yuwata commented on src/libsystemd-network/sd-dhcp6-client.c:
```diff
@@ -603,9 +612,15 @@ static int client_append_common_options_in_managed_mode(
-        r = dhcp6_option_append_fqdn(buf, offset, client->fqdn);
-        if (r < 0)
-                return r;
+        if (IN_SET(client->state,
+                      DHCP6_STATE_SOLICITATION,
+                      DHCP6_STATE_REQUEST,
+                      DHCP6_STATE_RENEW,
+                      DHCP6_STATE_REBIND)) {
```
Spurious indentation??

yuwata commented on src/libsystemd-network/sd-dhcp6-client.c:
```diff
@@ -679,6 +696,10 @@ static int client_append_oro(sd_dhcp6_client *client, uint8_t **buf, size_t *off
                 break;
 
+        case DHCP6_STATE_STOPPING:
+                n = 0;
```
Not a big matter, but please return earlier: `return 0;`

yuwata commented on src/libsystemd-network/test-dhcp6-client.c:
```diff
@@ -1010,6 +1081,7 @@ TEST(dhcp6_client) {
         assert_se(sd_dhcp6_client_set_local_address(client, &local_address) >= 0);
         assert_se(sd_dhcp6_client_set_fqdn(client, "host.lab.intra") >= 0);
         assert_se(sd_dhcp6_client_set_iaid(client, unaligned_read_be32((uint8_t[]) { IA_ID_BYTES })) >= 0);
+        assert_se(sd_dhcp6_client_set_send_release(client, 1) == 0);
```
This is a test code, hence not a big matter, but please use `true` even the argument type is `int`.

yuwata commented on src/network/networkd-dhcp6.c:
```diff
@@ -707,6 +707,12 @@ static int dhcp6_configure(Link *link) {
+        r = sd_dhcp6_client_set_send_release(client, link->network->dhcp6_send_release);
+        if (r < 0)
+                return log_link_debug_errno(link, r,
+                                            "DHCPv6 CLIENT: Failed to %s send release message on stop: %m",
```
`send` -> `sending`

yuwata reviewed (changes requested):
Please drop MUD URL from the release message, that is, conditionalize the call `dhcp6_option_append(..., SD_DHCP6_OPTION_MUD_URL_V6, ...)` in `dhcp6_client_send_message()`.

yuwata commented on src/libsystemd-network/sd-dhcp6-client.c:
```diff
@@ -1327,9 +1351,19 @@ static int client_receive_message(
         if (!client)
                 return 0;
 
+        if (client->send_release) {
```
I guess the condition should be improved.
- The client has lease,
- The lease has at lease one valid IA_NA or IA_PD.

yuwata commented on src/libsystemd-network/sd-dhcp6-client.c:
```diff
@@ -1327,9 +1351,19 @@ static int client_receive_message(
         if (!client)
                 return 0;
 
+        if (client->send_release) {
+                client_set_state(client, DHCP6_STATE_STOPPING);
+                r = dhcp6_client_send_message(client);
+                if (r < 0)
+                        return log_dhcp6_client_errno(client, r,
+                                                      "Failed to send to DHCP6 release message: %m");
+        }
```
Not a big matter, but I'd like to extract the logic into a separate function, e.g. `client_send_release()` or so.

yuwata commented on src/libsystemd-network/sd-dhcp6-client.c:
```diff
@@ -603,9 +612,15 @@ static int client_append_common_options_in_managed_mode(
+                   DHCP6_STATE_SOLICITATION,
+                   DHCP6_STATE_REQUEST,
+                   DHCP6_STATE_RENEW,
+                   DHCP6_STATE_REBIND)) {
```
This is OK, but for simplicity, how about `if (client->state != DHCP6_STATE_STOPPING)` ?

yuwata commented on src/libsystemd-network/sd-dhcp6-client.c:
```diff
@@ -690,6 +710,28 @@ static int client_append_oro(sd_dhcp6_client *client, uint8_t **buf, size_t *off
+                   DHCP6_STATE_SOLICITATION,
+                   DHCP6_STATE_REQUEST,
+                   DHCP6_STATE_RENEW,
+                   DHCP6_STATE_REBIND)) {
```
Also here. `if (client->state != DHCP6_STATE_STOPPING)` looks simpler.

yuwata commented on src/libsystemd-network/sd-dhcp6-client.c:
```diff
@@ -753,15 +795,14 @@ int dhcp6_client_send_message(sd_dhcp6_client *client) {
                 assert_not_reached();
         }
 
         if (client->mudurl) {
```
I'd like to move this condition into `client_append_mudurl()`.

yuwata commented on src/libsystemd-network/sd-dhcp6-client.c:
```diff
@@ -1319,17 +1362,38 @@ static int client_receive_message(
 }
 
+static void client_send_release(sd_dhcp6_client *client) {
+        int r;
+        sd_dhcp6_lease *lease;
```
Not a matter, and not a strong rule, but we usually write longer line first:
```
sd_dhcp6_lease *lease;
int r;
```
Also, missing assertion for `client`: `assert(client);`

yuwata commented on src/libsystemd-network/sd-dhcp6-client.c:
```diff
@@ -1319,17 +1362,38 @@ static int client_receive_message(
         if (!client)
                 return 0;
 
+        if (client->send_release)
```
Please move the condition into `client_send_release()`.

yuwata commented on src/libsystemd-network/sd-dhcp6-client.c:
```diff
@@ -1319,17 +1362,38 @@ static int client_receive_message(
+
+        client_set_state(client, DHCP6_STATE_STOPPING);
+        r = dhcp6_client_send_message(client);
+        if (r < 0)
+                log_dhcp6_client_errno(client, r,
+                                       "Failed to send to DHCP6 release message: %m");
```
Please propagate the failure to the caller, and ignore the failure in the caller side:
```c
static int client_send_release(sd_dhcp6_client *client) {
        ...;
        if (r < 0)
                return log_dhcp6_client_errno(client, r, "...");

        return 0;
}

int sd_dhcp6_client_stop(sd_dhcp6_client *client) {
        ...;

        (void) client_send_release(client);

        ...;
}
```

yuwata commented on src/libsystemd-network/test-dhcp6-client.c:
```diff
@@ -775,6 +831,15 @@ static void test_client_verify_solicit(const DHCP6Message *msg, size_t len) {
+        assert_se(len == sizeof(msg_release));
+        assert_se(msg->type == DHCP6_MESSAGE_RELEASE);
+        /* The transaction ID and elapsed time value are not deterministic. Skip them. */
+        assert_se(memcmp(msg->options, msg_release + offsetof(DHCP6Message, options), len - offsetof(DHCP6Message, options) - sizeof(be16_t)) == 0);
```
Not a big matter, but the line is slightly long. Please wrap it.

yuwata commented on src/libsystemd-network/sd-dhcp6-client.c:
```diff
@@ -690,6 +710,28 @@ static int client_append_oro(sd_dhcp6_client *client, uint8_t **buf, size_t *off
+        assert(*buf);
+        assert(offset);
+
+        int r;
```
Please declare variables before assertions:
```c
int r;

assert(client);
assert(buf);
...;
```

yuwata reviewed (changes requested):
Looks mostly good now. Several minor coding style comments.

poettering commented on NEWS:
```diff
@@ -55,6 +55,10 @@ CHANGES WITH 253 in spe:
 
+        * The [DHCPv6] section in .network file gained new SendRelease=
+          setting which enables the DHCPv6 client to send release when
+          it stops. This is the analog of the [DHCPv4] SendRelease= setting.
```
maybe also say this is enabled by default

Strangelovian commented:
@yuwata @poettering thanks for reviewing, I pushed a new update addressing your remarks

yuwata commented on src/libsystemd-network/sd-dhcp6-client.c:
```diff
@@ -690,6 +706,27 @@ static int client_append_oro(sd_dhcp6_client *client, uint8_t **buf, size_t *off
+        if (!client->mudurl)
+                return 0;
+
+        if (client->state != DHCP6_STATE_STOPPING) {
```
As @poettering suggested, please return earlier:
```c
if (client->state == DHCP6_STATE_STOPPING)
        return 0;

r = dhcp6_option_append(...);
```

Strangelovian replied:
yes fixed in the last push

yuwata commented on src/libsystemd-network/sd-dhcp6-client.c:
```diff
@@ -1319,17 +1355,44 @@ static int client_receive_message(
+        client_set_state(client, DHCP6_STATE_STOPPING);
+        r = dhcp6_client_send_message(client);
+        if (r < 0)
+                log_dhcp6_client_errno(client, r,
```
After reconsidering, please do not log the failure here, but log in the caller side, and mention that the failure is ignored:
```c
static int client_send_release(sd_dhcp6_client *client) {
        ...
        return dhcp6_client_send_message(client);
}

int sd_dhcp6_client_stop(sd_dhcp6_client *client) {
        ...
        r = client_send_release(client);
        if (r < 0)
                log_dhcp6_client_errno(client, r, "Failed to send to DHCP6 release message, ignoring: %m");
        ...
}
```

Strangelovian replied:
yes fixed in the last push

yuwata reviewed (changes requested):
Mostly good. Two super minor comments.

bluca commented:
needs a rebase

Strangelovian commented:
> needs a rebase

Yes, pushed

Strangelovian commented:
@yuwata @bluca sorry about the labeler bot, it crashed and the label are wrong. The rebase has been done, and some reviewing is needed

yuwata commented on src/libsystemd-network/sd-dhcp6-client.c:
```diff
@@ -1319,17 +1350,45 @@ static int client_receive_message(
+        r = client_send_release(client);
+        if (r < 0)
+                log_dhcp6_client_errno(client, r,
+                                       "Failed to send to DHCP6 release message: %m");
```
Please explicitly mention that the failure is ignored. That is, append ", ignoring" to the log message:
`"Failed to send to DHCP6 release message, ignoring: %m"`

Strangelovian replied:
Oh sorry, in the log message: done in the last push

yuwata reviewed (approved):
LGTM. Thank you!

Strangelovian commented:
> LGTM. Thank you!

@yuwata Thanks!
@poettering CI should be OK, can you check it too?
